cmd/cosign/cli/copy/copy.go

```diff
@@ -16,65 +16,75 @@ package copy
 	"context"
+	"errors"
 	"fmt"
+	"net/http"
 	"os"
 
 	"github.com/google/go-containerregistry/pkg/name"
 	v1 "github.com/google/go-containerregistry/pkg/v1"
 	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
 	"github.com/sigstore/cosign/cmd/cosign/cli/options"
+	"github.com/sigstore/cosign/pkg/oci"
 	ociremote "github.com/sigstore/cosign/pkg/oci/remote"
+	"github.com/sigstore/cosign/pkg/oci/walk"
 )
 
 // CopyCmd implements the logic to copy the supplied container image and signatures.
 // nolint
 func CopyCmd(ctx context.Context, regOpts options.RegistryOptions, srcImg, dstImg string, sigOnly, force bool) error {
-	var refs []name.Reference // nolint: staticcheck
 	srcRef, err := name.ParseReference(srcImg)
 	if err != nil {
 		return err
 	}
-
-	refs = append(refs, srcRef)
+	srcRepoRef := srcRef.Context()
 
 	dstRef, err := name.ParseReference(dstImg)
 	if err != nil {
 		return err
 	}
+	dstRepoRef := dstRef.Context()
 
 	remoteOpts := regOpts.GetRegistryClientOpts(ctx)
-	sigSrcRef, err := ociremote.SignatureTag(srcRef, ociremote.WithRemoteOptions(remoteOpts...))
+	root, err := ociremote.SignedEntity(srcRef, ociremote.WithRemoteOptions(remoteOpts...))
 	if err != nil {
 		return err
 	}
 
-	dstRepoRef := dstRef.Context()
-	sigDstRef := dstRepoRef.Tag(sigSrcRef.Identifier())
-	if err := copyImage(sigSrcRef, sigDstRef, force, remoteOpts...); err != nil {
-		return err
-	}
-
-	if !sigOnly {
-		attSrcRef, err := ociremote.AttestationTag(srcRef, ociremote.WithRemoteOptions(remoteOpts...))
+	return walk.SignedEntity(ctx, root, func(ctx context.Context, se oci.SignedEntity) error {
+		// Both of the SignedEntity types implement Digest()
+		h, err := se.(interface{ Digest() (v1.Hash, error) }).Digest()
 		if err != nil {
 			return err
 		}
-		refs = append(refs, attSrcRef)
+		srcDigest := srcRepoRef.Digest(h.String())
 
-		sbomSrcRef, err := ociremote.SBOMTag(srcRef, ociremote.WithRemoteOptions(remoteOpts...))
-		if err != nil {
+		// Copy signatures.
+		if err := copyTagImage(ociremote.SignatureTag, srcDigest, dstRepoRef, force, remoteOpts...); err != nil {
 			return err
 		}
-		refs = append(refs, sbomSrcRef)
+		if sigOnly {
+			return nil
+		}
 
-		for _, ref := range refs {
-			if err := copyImage(ref, dstRepoRef.Tag(ref.Identifier()), force, remoteOpts...); err != nil {
-				fmt.Fprintf(os.Stderr, "WARNING: %s tag could not be found for an image %s, it might not exist, continuing anyway\n", ref.Identifier(), srcImg)
-			}
+		// Copy attestations
+		if err := copyTagImage(ociremote.AttestationTag, srcDigest, dstRepoRef, force, remoteOpts...); err != nil {
+			return err
+		}
+
+		// Copy SBOMs
+		if err := copyTagImage(ociremote.SBOMTag, srcDigest, dstRepoRef, force, remoteOpts...); err != nil {
+			return err
+		}
+
+		// Copy the entity itself.
+		if err := copyImage(srcDigest, dstRepoRef.Tag(srcDigest.Identifier()), force, remoteOpts...); err != nil {
+			return err
```

Comment:
I think there's a subtle bug here where the image itself is copied by digest and the tag isn't necessarily updated.  Lemme update this.

Reply:
Ok, this should be resolve.